File: 36-valid-sudoku/valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // Check for valid rows
        for(int i=0; i<9; i++){
            unordered_set<int> st;
            for(int j=0; j<9; j++){
                if (board[i][j] <= '9' && board[i][j] > '0'){
                    int x = board[i][j] - '0';
                    if (st.count(x)) return false;
                    st.insert(x);
                }
            }
        }
        // Check for valid cols
        for(int i=0; i<9; i++){
            unordered_set<int> st;
            for(int j=0; j<9; j++){
                if (board[j][i] <= '9' && board[j][i] > '0'){
                    int x = board[j][i] - '0';
                    if (st.count(x)) return false;
                    st.insert(x);
                }
            }
        }
        // Check for those 3x3 squares
        for(int l=0; l<9; l+=3){
            for(int k=0; k<9; k+=3){
                unordered_set<int> st;
                for(int i=l; i<3+l; i++){
                    for(int j=k; j<3+k; j++){
                        if (board[i][j] <= '9' && board[i][j] > '0'){
                            int x = board[i][j] - '0';
                            if (st.count(x)) return false;
                            st.insert(x);
                        }
                    }
                }
            }
        }
        return true;
    }
};

File: 36-valid-sudoku/valid-sudoku.go
func isValidSudoku(board [][]byte) bool {
    // check all the rows
    for i := 0; i < 9; i++ {
        seen := map[byte]bool{}
        for j := 0; j < 9; j++ {
            if board[i][j] == '.' {
                continue
            }
            if _, ok := seen[board[i][j]]; ok {
                return false
            }
            seen[board[i][j]] = true
        }
    }
    // check all the columns
    for i := 0; i < 9; i++ {
        seen := map[byte]bool{}
        for j := 0; j < 9; j++ {
            if board[j][i] == '.' {
                continue
            }
            if _, ok := seen[board[j][i]]; ok {
                return false
            }
            seen[board[j][i]] = true
        }
    }
    // check all 3x3 submatrices
    for row := 0; row < 9; row += 3 {
        for col := 0; col < 9; col += 3 {
            seen := map[byte]bool{}
            for i := row; i < row + 3; i++ {
                for j := col; j < col + 3; j++ {
                    if board[i][j] == '.' {
                        continue
                    }
                    if _, ok := seen[board[i][j]]; ok {
                        return false
                    }
                    seen[board[i][j]] = true
                }
            }
        }
    }
    return true
}